launcher/minecraft/launch/VerifyJavaInstall.cpp

```diff
@@ -1,50 +1,75 @@
-#include <minecraft/VersionFilterData.h>
+#include "VerifyJavaInstall.h"
 
-#ifdef major
-    #undef major
-#endif
-#ifdef minor
-    #undef minor
-#endif
+#include "java/JavaVersion.h"
+#include "minecraft/PackProfile.h"
+#include "minecraft/MinecraftInstance.h"
 
 void VerifyJavaInstall::executeTask() {
-    auto m_inst = std::dynamic_pointer_cast<MinecraftInstance>(m_parent->instance());
-
-    auto javaVersion = m_inst->getJavaVersion();
-    auto minecraftComponent = m_inst->getPackProfile()->getComponent("net.minecraft");
-
-    // Java 17 requirement
-    if (minecraftComponent->getReleaseDateTime() >= g_VersionFilterData.java17BeginsDate) {
-        if (javaVersion.major() < 17) {
-            emit logLine("Minecraft 1.18 Pre Release 2 and above require the use of Java 17",
-                         MessageLevel::Fatal);
-            emitFailed(tr("Minecraft 1.18 Pre Release 2 and above require the use of Java 17"));
-            return;
-        }
-    }
-    // Java 16 requirement
-    else if (minecraftComponent->getReleaseDateTime() >= g_VersionFilterData.java16BeginsDate) {
-        if (javaVersion.major() < 16) {
-            emit logLine("Minecraft 21w19a and above require the use of Java 16",
-                         MessageLevel::Fatal);
-            emitFailed(tr("Minecraft 21w19a and above require the use of Java 16"));
-            return;
-        }
+    auto instance = std::dynamic_pointer_cast<MinecraftInstance>(m_parent->instance());
+    auto packProfile = instance->getPackProfile();
+    auto settings = instance->settings();
+    auto storedVersion = settings->get("JavaVersion").toString();
+    auto ignoreCompatibility = settings->get("IgnoreJavaCompatibility").toBool();
+
+    auto compatibleMajors = packProfile->getProfile()->getCompatibleJavaMajors();
+
+    JavaVersion javaVersion(storedVersion);
+
+    if (compatibleMajors.isEmpty() || compatibleMajors.contains(javaVersion.major()))
+    {
+        emitSucceeded();
+        return;
     }
-    // Java 8 requirement
-    else if (minecraftComponent->getReleaseDateTime() >= g_VersionFilterData.java8BeginsDate) {
-        if (javaVersion.major() < 8) {
-            emit logLine("Minecraft 17w13a and above require the use of Java 8",
-                         MessageLevel::Fatal);
-            emitFailed(tr("Minecraft 17w13a and above require the use of Java 8"));
-            return;
-        }
+
+
+    if (ignoreCompatibility)
+    {
+        emit logLine(tr("Java major version is incompatible. Things might break."), MessageLevel::Warning);
+        emitSucceeded();
+        return;
     }
 
-    emitSucceeded();
+    emit logLine(tr("Instance not compatible with Java major version %1.\n"
+                    "Switch the Java version of this instance to one of the following:").arg(javaVersion.major()),
+                 MessageLevel::Error);
+    for (auto major: compatibleMajors)
+    {
+        emit logLine(tr("Java %1").arg(major), MessageLevel::Error);
+    }
```

Comment:
Not happy with this. Maybe we can display it differently?

Meta can suggest multiple versions. For example 1.17-pre1 will recommend `16` and `17`

Reply:
> Meta can suggest multiple versions. For example 1.17-pre1 will recommend `16` and `17`

I personally wouldn't have it suggest Java 16, since it is discontinued and no longer supported. Security and all that.

Reply:
It's suggesting Java 16, because that's what Mojang is suggesting. In any case this is handled by meta, so if we want to force newer Javas in the future we can totally do that.